# Кондратюк Андрій       КМ-62        Варіант 10

## <center>Завдання на екзамен</center>
1.	Файл з даними містить: час вимірювання, перший сигнал, другий сигнал. Побудувати графіки сигналів.
2.	Знайти основні статистичні характеристики: середнє, дисперсію, гістограму розподілу, кореляційну функцію, спектральну щільність (перетворення Фур’є).
3.	Перевірити кореляцію між сигналами. Якщо кореляція існує, то побудувати лінійну залежність між сигналами (перший сигнал –y, другий сигнал -x). Якщо кореляція відсутня, то побудувати квадратичну залежність. Оцінити адекватність моделей та значимість коефіцієнтів.
4.	Побудувати вейвлет – перетворення для двох сигналів. Порівняти результати. (Взяти один з вейвлетів – гауссовий, Шенона, Хаара, на вибір).
5.	Провести однофакторний аналіз (сигнали – це два рівні деякого фактора).
6.	Провести аналіз отриманих результатів.

## <center>Завдання №1</center>

#### Першим кроком буде завантаження даних

In [9]:
from csv import reader

opened_file = open(r"C:\Users\Andrey\Downloads\V10.txt")
read_file = reader(opened_file)
data_set = list(read_file)
del(data_set[-1])

#### Виведемо перших п'ять та останніх рядків. Визначимо загальну кількість рядків

In [10]:
def explore_data(data_set):
    data_set_slice = data_set[:5] + data_set[-5:]
    for i in range(len(data_set_slice)):
        if i == 0:
            print("Перші п'ять рядків\n")
        if i == 5:
            print("\n...\n\nОстанні 5 рядків")
        
        print(f"Час: {data_set_slice[i][0]}\tПерший Сигнал: {data_set_slice[i][1]}\tДругий Сигнал: {data_set_slice[i][2]}")
        
    print(f"\nЗагальна кількість рядків: {len(data_set)}")
    print(f"Кількість стовбців: {len(data_set[0])}")

explore_data(data_set)

Перші п'ять рядків

Час: 0.000	Перший Сигнал: 0.175	Другий Сигнал: 0.635
Час: 0.003	Перший Сигнал: 0.175	Другий Сигнал: 0.635
Час: 0.006	Перший Сигнал: 0.175	Другий Сигнал: 0.635
Час: 0.008	Перший Сигнал: 0.175	Другий Сигнал: 0.635
Час: 0.011	Перший Сигнал: 0.175	Другий Сигнал: 0.635

...

Останні 5 рядків
Час: 59.986	Перший Сигнал: -0.160	Другий Сигнал: 0.120
Час: 59.989	Перший Сигнал: -0.210	Другий Сигнал: 0.115
Час: 59.992	Перший Сигнал: -0.230	Другий Сигнал: 0.130
Час: 59.994	Перший Сигнал: -0.240	Другий Сигнал: 0.135
Час: 59.997	Перший Сигнал: -0.255	Другий Сигнал: 0.135

Загальна кількість рядків: 21600
Кількість стовбців: 3


#### Побудуємо три графіки

In [11]:
import plotly.plotly as py
import plotly.graph_objs as go

time          = [float(row[0]) for row in data_set] 
first_signal  = [float(row[1]) for row in data_set] 
second_signal = [float(row[2]) for row in data_set]
    
trace_1 = go.Scatter(
    x = time,
    y = first_signal,
    name = "Перший Сигнал",
    line = dict(color = '#17BECF'),
    opacity = 0.8)

trace_2 = go.Scatter(
    x = time,
    y = second_signal,
    name = "Другий Сигнал",
    line = dict(color = '#7F7F7F'),
    opacity = 0.8)

data = [trace_1, trace_2]
    
layout_1 = dict(
    title='Значення першого сигналу в залежності від часу',
    xaxis = dict(
        rangeslider=dict(
            visible = True
        ),  title = "Час"
    ),
    yaxis = dict(title = "Значення")
)
layout_2 = dict(
    title='Значення другого сигналу в залежності від часу',
    xaxis = dict(
        rangeslider=dict(
            visible = True
        ),  title = "Час"
    ),
    yaxis = dict(title = "Значення")
)
layout_3 = dict(
    title='Значення сигналів в залежності від часу',
    xaxis = dict(
        rangeslider=dict(
            visible = True
        ),  title = "Час"
    ),
    yaxis = dict(title = "Значення")
)

fig_1 = dict(data=[trace_1], layout=layout_1)
fig_2 = dict(data=[trace_2], layout=layout_2)
fig_3 = dict(data=data, layout=layout_3)

In [27]:
py.iplot(fig_1, filename = "Значення першого сигналу в залежності від часу")

In [13]:
py.iplot(fig_2, filename = "Значення другого сигналу в залежності від часу")

In [26]:
py.iplot(fig_3, filename = "Значення сигналів в залежності від часу")

The draw time for this plot will be slow for clients without much RAM.


#### Висновоки 
    На графіках можна спостерігати високу частоту сигналів та їх періодичність.

## <center>Завдання №2</center>

#### Середнє для сигналів

In [15]:
both_signal = first_signal + second_signal #concatenate two signals in one list 
avg_first_signal = sum(first_signal) / len(first_signal)
avg_second_signal = sum(second_signal) / len(second_signal)
avg_both_signal = sum(both_signal) / len(both_signal)
print(f"Cеререднє значення для:\nПершого сигналу = {round(avg_first_signal,6)}\nДругого сигналу =  {round(avg_second_signal,6)}")
print(f"Обох сигналів =   {round(avg_both_signal,6)}")

Cеререднє значення для:
Першого сигналу = -0.278269
Другого сигналу =  0.063986
Обох сигналів =   -0.107142


#### Дисперсія для сигналів

In [16]:
var_first_signal = sum((x_i - avg_first_signal) ** 2 for x_i in first_signal) / len(first_signal)
var_second_signal = sum((x_i - avg_first_signal) ** 2 for x_i in second_signal) / len(second_signal)
var_both_signal = sum((x_i - avg_both_signal) ** 2 for x_i in both_signal) / len(both_signal)
print(f"Значення дисперсії для:\nПершого сигналу =  {round(var_first_signal,6)}\nДругого сигналу =  {round(var_second_signal,6)}")
print(f"Обох сигналів   =  {round(var_both_signal,6)}")

Значення дисперсії для:
Першого сигналу =  0.172778
Другого сигналу =  0.24261
Обох сигналів   =  0.178409


#### Побудова гістограми розподілу

In [25]:
import plotly.figure_factory as ff

hist_data = [first_signal, second_signal]

group_labels = ['Перший сигнал', 'Другий сигнал']

colors = ['#17BECF', '#7F7F7F']

fig = ff.create_distplot(hist_data, group_labels, bin_size=.16, curve_type='normal', colors=colors)

fig['layout'].update(title='Гістограма двох сигналів з нормальним розподілом')

py.iplot(fig, filename='Гістограма двох сигналів з нормальним розподілом')

The draw time for this plot will be slow for clients without much RAM.


#### Висновки
    - Перший сигнал "перекошений вправо", означає, що велика частка даних розподілена з лівої сторони, а довгий "хвіст" 
    даних простягається вправо
    - Другий сигнал "перекошений вліво", тобто його хвіст зміщений в ліву сторону. Що означає, більша частина даних  
    сконцентрована за правою стороною гаусівського розподілу

#### Кореляційна функція

In [19]:
from scipy import signal
import numpy as np

y_1 = first_signal
y_2 = second_signal
n = len(first_signal)
ry_1 = np.correlate(y_1, y_1, mode = 'same') / n
ry_2 = np.correlate(y_2, y_2, mode = 'same') / n
l = np.arange(-n / 2, n / 2)


trace_1 = go.Scatter(
    x = l,
    y = ry_1,
    name = "Перший Сигнал",
    line = dict(color = '#17BECF'),
    opacity = 0.8)

trace_2 = go.Scatter(
    x = l,
    y = ry_2,
    name = "Другий Сигнал",
    line = dict(color = '#7F7F7F'),
    opacity = 0.8)

layout = dict(
    title = "Кореаліаційна функція першого та другого сигналу"
)

data = [trace_1, trace_2]

fig = dict(data=data, layout=layout)
py.iplot(fig, filename='Кореаліаційна функція першого та другого сигналу')

The draw time for this plot will be slow for clients without much RAM.


АКФ застосовується для знаходження періодичних сигналів, не видимих через додатковий шум 

#### Спектральна щільність

In [20]:
from scipy import signal

freqs_1, psd_1 = signal.welch(first_signal)
freqs_2, psd_2 = signal.welch(second_signal)

trace_1 = go.Scatter(
    x = freqs_1,
    y = psd_1,
    name = "Перший Сигнал",
    line = dict(color = '#17BECF'),
    opacity = 0.8)

trace_2 = go.Scatter(
    x = freqs_2,
    y = psd_2,
    name = "Другий Сигнал",
    line = dict(color = '#7F7F7F'),
    opacity = 0.8)

layout = dict(
    title = "Cпектральна щільність першого та другого сигналу",
    xaxis = dict(
        type = 'log',
        autorange = True)
)

data = [trace_1, trace_2]

fig = dict(data=data, layout=layout)
py.iplot(fig, filename='Cпектральна щільність першого та другого сигналу')

Даний графік зображує кількісну потужність сигналу для кожної частоти

## <center>Завдання №3</center>

#### Перевірка залежності між сигналами

In [21]:
corr_coef = np.corrcoef(first_signal, second_signal)
print(f"Коефіцієнт кореляції між першим та другим сигналом = {corr_coef[0][1]}")

Коефіцієнт кореляції між першим та другим сигналом = -0.5401768627527949


#### Побудова графіків залежності

In [22]:
x = second_signal
y = first_signal

z = np.polyfit(x, y, 1)
f = np.poly1d(z)
print(f)

x_new = np.linspace(-1.5, 1.5, 5000)
y_new = f(x_new)

trace_1 = go.Scatter(
    x = x,
    y = y,
    mode = 'markers',
    name = 'Сигнали',
    opacity = 0.8)

trace_2 = go.Scatter(
    x = x_new,
    y = y_new,
    mode = 'lines',
    name = 'Лінійна функція'
)

layout = dict(
    title='Лінійна залежність між сигналами'
)

data = [trace_1, trace_2]

fig = dict(data=data, layout=layout)
py.iplot(fig, filename='basic-scatter')

 
-0.6339 x - 0.2377


In [64]:
x = second_signal
y = first_signal

z = np.polyfit(x, y, 2)
f = np.poly1d(z)
print(f)

x_new = np.linspace(-1.5, 1.5, 5000)
y_new = f(x_new)

trace1 = go.Scatter(
    x=x,
    y=y,
    mode='markers',
    name='Сигнали'
)

trace2 = go.Scatter(
    x=x_new,
    y=y_new,
    mode='lines',
    name='Квадратична функція'
)

layout = dict(
    title='Квадратична залежність між сигналами',
)

data = [trace1, trace2]

fig = dict(data=data, layout=layout)
py.iplot(data, filename='Polynomial-Fit-in-python')

       2
0.431 x - 0.4711 x - 0.304


#### Висновки.
    Отримано негативний коефіцієнт кореляції між двома сигналами: -0.5401768627527949. Більш менш, слабка кореляція 
    між параметрами, тому опиратись на модель лінійної залежності не варто. Більш адекватною є модель 
    квадратичної залежності

## <center>Завдання №4</center>

#### Вейвлет перетворення

In [63]:
def lowpassfilter(signal, thresh = 0.63, wavelet="haar"):
    thresh = thresh * np.nanmax(signal)
    coeff = pywt.wavedec(signal, wavelet, mode="per" )
    coeff[1:] = (pywt.threshold(i, value=thresh, mode="soft" ) for i in coeff[1:])
    reconstructed_signal = pywt.waverec(coeff, wavelet, mode="per" )
    return reconstructed_signal

rec = lowpassfilter(first_signal, 0.4)

trace1 = go.Scatter(
    x=time,
    y=first_signal,
    mode='lines',
    name='Перший сигнал',
    line = dict(color = '#17BECF'),
    opacity = 0.7
)

trace2 = go.Scatter(
    x=time,
    y=rec,
    mode='lines',
    name='Перетворений перший сигнал'
)

layout = dict(
    title='Вейвлет перетворення Хаара для пешого сигналу',
)

data = [trace1, trace2]

fig = dict(data=data, layout=layout)
py.iplot(data, filename='Вейвлет перетворення Хаара для пешого сигналу')

The draw time for this plot will be slow for clients without much RAM.


In [62]:
rec = lowpassfilter(second_signal, 0.4)

trace1 = go.Scatter(
    x=time,
    y=first_signal,
    mode='lines',
    name='Другий сигнал',
    line = dict(color = '#7F7F7F'),
    opacity = 0.7
)

trace2 = go.Scatter(
    x=time,
    y=rec,
    mode='lines',
    name='Перетворений другий сигнал'
)

layout = dict(
    title='Вейвлет перетворення Хаара для другого сигналу',
)

data = [trace1, trace2]

fig = dict(data=data, layout=layout)
py.iplot(data, filename='Вейвлет перетворення Хаара для другого сигналу')

The draw time for this plot will be slow for clients without much RAM.


#### Висновки
    Застосовано вейвлет перетворення Хаарана. Як ми можемо бачити, шляхом деконструкції сигналу, встановлення деяких   
    коефіцієнтів до нуля, і відновлення його знову, ми можемо видалити високочастотний шум від сигналу

## <center>Завдання №5</center>

#### Однофакторний аналіз, де сигнали – це два рівні деякого фактора

In [74]:
def FactorAnalis(x, y):
    factors = 2
    n = len(y)
    Si2 = [0,0]
    xis = []
    list = [[i for i in x], [j for j in y]]

    xis.append(np.mean(x))
    xis.append(np.mean(y))

    for i in range(factors):
        for j in range(n):
            Si2[i] += ((list[i][j] - xis[i]) ** 2)
        Si2[i] /= (n - 1)

    g = max(Si2) / sum(Si2)

    print('g = max(si)/sum(si) = ', g)
    print('g_alf: F_0.95_(3,1000) = 2.6')
    print('нульова гіпотеза виконується g < g_alf')

    S02 = 0
    st_r = 0
    temp = 0
    st_l = 0
    for i in range(factors):
        for j in range(n):
            st_l += list[i][j] ** 2
            temp += list[i][j]
        st_r += temp ** 2
        temp = 0

    S02  = 1 / (factors * (n - 1)) * (st_l - st_r / n)
    print('Оцінка дисперсії, що характеризується рзсіюванням поза впливом фактора: ', S02)

    S2 = 0
    st_r = 0
    st_l = 0
    xi = [0 for row in range(factors)]
    temp = 0
    x_ss = 0
    for i in range(factors):
        for j in range(n):
            st_l += list[i][j] ** 2
            st_r += list[i][j]
    S2 = 1 / (factors * n - 1) * (st_l - 1 / (factors * n) * st_r ** 2)
    print('Вибіркова дисперсія всіх спостережень: ', S2)

    for i in range(factors):
        for j in range(n):
            temp += list[i][j]
        xi[i] = temp / n
        temp = 0

    for i in range(factors):
        x_ss += xi[i]
    x_ss /= factors

    Sa2 = 0

    for i in range(factors):
        Sa2 += (xi[i] - x_ss) ** 2
    Sa2 = Sa2 * n / (factors - 1)
    print("Оцінка дисперсії, що характеризує зміни параметра, повязанні з фактором: ", Sa2)
    print("Оцінка впливу фактора на зміни середнього значення: ")
    print("Sa/S: ", Sa2 / S2)
    print("F0.95(f1,f2), f1 - 2, f2 - 21600: 3")
    
FactorAnalis(first_signal, second_signal)

g = max(si)/sum(si) =  0.5793075753557919
g_alf: F_0.95_(3,1000) = 2.6
нульова гіпотеза виконується g < g_alf
Оцінка дисперсії, що характеризується рзсіюванням поза впливом фактора:  0.1491314809796667
Вибіркова дисперсія всіх спостережень:  0.1784133833981288
Оцінка дисперсії, що характеризує зміни параметра, повязанні з фактором:  1265.0980340561287
Оцінка впливу фактора на зміни середнього значення: 
Sa/S:  7090.824746219107
F0.95(f1,f2), f1 - 2, f2 - 21600: 3


#### Висновки
    Отже, можна сказати, що існує вплив

## <center>Висновки</center>

    Проаналізувавши обидва сигнали, можна сказати що вони мають слабку негативну кореляцію. Практично нормально розподілені,
    мають хвости, де більша частина даних сконцентрована. Сигнали мають високі частоти, які можна прибрати за допомогою 
    вейвлет перетворення.